src/smt/theory_str_noodler/theory_str_noodler.cpp

```diff
@@ -1889,6 +1893,101 @@ namespace smt::noodler {
         }
     }
 
+    /**
+     * @brief Handle str.<=
+     * Translates to the following axiom
+     * 
+     * x <= y -> x = y | x < y
+     * @param e str.<= predicate
+     */
+    void theory_str_noodler::handle_lex_leq(expr *e) {
+        if(axiomatized_persist_terms.contains(e))
+            return;
+
+        axiomatized_persist_terms.insert(e);
+        STRACE("str", tout  << "handle str.<= " << mk_pp(e, m) << std::endl;);
+
+        expr *x = nullptr, *y = nullptr;
+        VERIFY(m_util_s.str.is_le(e, x, y));
+      
+        expr_ref e_lt(m_util_s.str.mk_lex_lt(x, y), m);
+        expr_ref x_y(m.mk_eq(x,y), m);
+        literal lit_e_lt = mk_literal(e_lt);
+        literal lit_e = mk_literal(e);
+        literal lit_x_y = mk_literal(x_y);
+        // x <= y -> x = y | x < y
+        add_axiom({~lit_e, lit_e_lt, lit_x_y});
+    }
+
+    /**
+     * @brief Handle str.<
+     * Translates to the following theory axioms.
+     * 
+     * x = eps & y != eps -> x < y
+     * x < y & x != eps -> x = u.v1.w1
+     * x < y & x != eps -> v1 in re.allchar
+     * x < y & x != eps -> v2 in re.allchar
+     * x < y & x != eps -> to_code(v1) + k = to_code(v2) & k >= 1
```

Comment:
This is overly convoluted, you can just do `to_code(v1) < to_code(v2)`.

Reply:
Not really. There is some sad story inside the function.

Reply:
Well, that is weird. Maybe it becomes nonlinear because `to_code` is a function? But still, does not make sense if your hack works.

Reply:
Yes, thats strange.